File: scripts/create_collector_creditor.py
#!/usr/bin/env python3
"""
Script to create collector and creditor accounts in the Supabase database.
This script creates user accounts with the appropriate roles and profiles.
"""

import os
import requests
import json
import uuid
from datetime import datetime

# Load environment variables
from dotenv import load_dotenv
load_dotenv()

# Supabase configuration
SUPABASE_URL = os.getenv('VITE_SUPABASE_URL')
SERVICE_ROLE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

if not SUPABASE_URL or not SERVICE_ROLE_KEY:
    print("Missing Supabase environment variables. Please check your .env file.")
    print("Required variables: VITE_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY")
    exit(1)

# Headers for Supabase requests
headers_admin = {
    'apikey': SERVICE_ROLE_KEY,
    'Authorization': f'Bearer {SERVICE_ROLE_KEY}',
    'Content-Type': 'application/json',
}

def create_user(email, password, full_name):
    """Create a user account in Supabase Auth"""
    print(f"Creating user account for {email}...")
    
    # Create user via Supabase Auth API
    signup_url = f"{SUPABASE_URL}/auth/v1/signup"
    signup_payload = {
        "email": email,
        "password": password,
        "data": {
            "full_name": full_name
        }
    }
    
    response = requests.post(signup_url, headers=headers_admin, json=signup_payload)
    
    if response.status_code not in [200, 201]:
        print(f"Error creating user: {response.status_code} - {response.text}")
        return None
        
    user_data = response.json()
    user_id = user_data.get('user', {}).get('id')
    print(f"User created with ID: {user_id}")
    return user_id

def create_profile(user_id, email, full_name):
    """Create a profile record for the user"""
    print(f"Creating profile for user {user_id}...")
    
    profile_data = {
        "id": user_id,
        "full_name": full_name,
        "email": email,
        "updated_at": datetime.utcnow().isoformat()
    }
    
    # Upsert profile
    profile_url = f"{SUPABASE_URL}/rest/v1/profiles"
    response = requests.post(profile_url, headers=headers_admin, json=profile_data)
    
    if response.status_code not in [200, 201]:
        print(f"Error creating profile: {response.status_code} - {response.text}")
        return False
        
    print("Profile created successfully")
    return True

def create_staff_record(user_id, role_name):
    """Create a staff record for collector or creditor"""
    print(f"Creating {role_name} record for user {user_id}...")
    
    # Generate unique employee ID
    employee_id = f"{role_name.upper()}{uuid.uuid4().hex[:6].upper()}"
    
    staff_data = {
        "user_id": user_id,
        "employee_id": employee_id
    }
    
    # Insert staff record
    staff_url = f"{SUPABASE_URL}/rest/v1/staff"
    response = requests.post(staff_url, headers=headers_admin, json=staff_data)
    
    if response.status_code not in [200, 201]:
        print(f"Error creating staff record: {response.status_code} - {response.text}")
        return False
        
    print(f"{role_name.capitalize()} record created successfully")
    return True

def assign_role(user_id, role):
    """Assign a role to the user"""
    print(f"Assigning {role} role to user {user_id}...")
    
    role_data = {
        "user_id": user_id,
        "role": role,
        "active": True
    }
    
    # Insert role assignment
    role_url = f"{SUPABASE_URL}/rest/v1/user_roles"
    response = requests.post(role_url, headers=headers_admin, json=role_data)
    
    if response.status_code not in [200, 201]:
        print(f"Error assigning role: {response.status_code} - {response.text}")
        return False
        
    print(f"Role {role} assigned successfully")
    return True

def create_collector(email, full_name, password="Test1234!"):
    """Create a collector account"""
    print(f"\n=== Creating Collector: {full_name} ===")
    
    # Create user account
    user_id = create_user(email, password, full_name)
    if not user_id:
        return False
    
    # Create profile
    if not create_profile(user_id, email, full_name):
        return False
    
    # Create staff record
    if not create_staff_record(user_id, "collector"):
        return False
    
    # Assign collector role
    if not assign_role(user_id, "collector"):
        return False
    
    print(f"Collector {full_name} created successfully!")
    print(f"  Email: {email}")
    print(f"  Password: {password}")
    return True

def create_creditor(email, full_name, password="Test1234!"):
    """Create a creditor account"""
    print(f"\n=== Creating Creditor: {full_name} ===")
    
    # Create user account
    user_id = create_user(email, password, full_name)
    if not user_id:
        return False
    
    # Create profile
    if not create_profile(user_id, email, full_name):
        return False
    
    # Create staff record
    if not create_staff_record(user_id, "creditor"):
        return False
    
    # Assign creditor role
    if not assign_role(user_id, "creditor"):
        return False
    
    print(f"Creditor {full_name} created successfully!")
    print(f"  Email: {email}")
    print(f"  Password: {password}")
    return True

def main():
    print("=== Collector and Creditor Account Creation Script ===")
    print(f"Supabase URL: {SUPABASE_URL}")
    
    # Get user input for collector account
    collector_email = input("Enter collector email (or press Enter for default): ").strip()
    if not collector_email:
        collector_email = "collector@example.com"
    
    collector_name = input("Enter collector full name (or press Enter for default): ").strip()
    if not collector_name:
        collector_name = "John Collector"
    
    # Get user input for creditor account
    creditor_email = input("Enter creditor email (or press Enter for default): ").strip()
    if not creditor_email:
        creditor_email = "creditor@example.com"
    
    creditor_name = input("Enter creditor full name (or press Enter for default): ").strip()
    if not creditor_name:
        creditor_name = "Jane Creditor"
    
    # Create collector
    create_collector(
        email=collector_email,
        full_name=collector_name,
        password="Test1234!"
    )
    
    # Create creditor
    create_creditor(
        email=creditor_email,
        full_name=creditor_name,
        password="Test1234!"
    )
    
    print("\n=== Account Creation Complete ===")
    print("You can now log in to the system using these accounts.")

if __name__ == "__main__":
    main()

File: scripts/createCollectorCreditor.ts
// Script to create collector and creditor accounts
import { createClient } from '@supabase/supabase-js';
import * as dotenv from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';

// Get __dirname equivalent for ES modules
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load environment variables
dotenv.config({ path: resolve(__dirname, '../.env') });

// Initialize Supabase client with the correct environment variables
const supabaseUrl = process.env.VITE_SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error('Missing Supabase environment variables. Please check your .env file.');
  console.error('Required variables: VITE_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY');
  process.exit(1);
}

console.log('Connecting to Supabase at:', supabaseUrl);

const supabase = createClient(supabaseUrl, supabaseKey);

// Password for all accounts (at least 6 characters for Supabase)
const PASSWORD = 'Test1234!';

// Utility function to add delay
function delay(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function createCollectorAccount(email: string, fullName: string) {
  console.log(`Creating collector account for ${fullName}...`);
  
  // First, create the user account
  const { data: authData, error: authError } = await supabase.auth.signUp({
    email: email,
    password: PASSWORD,
    options: {
      data: {
        full_name: fullName,
      }
    }
  });

  if (authError) {
    console.error('Error creating collector auth account:', authError);
    return null;
  }

  if (!authData || !authData.user) {
    console.error('Failed to create collector auth account - no user data returned');
    return null;
  }

  console.log('Collector auth account created:', authData.user.id);
  
  // Create profile record first
  const profileData = {
    id: authData.user.id,
    full_name: fullName,
    email: email,
    updated_at: new Date().toISOString(),
  };

  const { error: profileError } = await supabase
    .from('profiles')
    .upsert([profileData], { onConflict: 'id' });

  if (profileError) {
    console.error('Error creating profile record:', profileError);
    return null;
  }

  // Generate unique employee ID
  const employeeId = `COLLECTOR${Math.floor(Math.random() * 10000)}`;
  
  // Create staff record
  const staffData = {
    user_id: authData.user.id,
    employee_id: employeeId,
  };

  const { data: staffDataResult, error: staffError } = await supabase
    .from('staff')
    .insert([staffData])
    .select();

  if (staffError) {
    console.error('Error creating staff record:', staffError);
    return null;
  }

  // Assign collector role
  const userRoleData = {
    user_id: authData.user.id,
    role: 'collector',
    active: true,
  };

  const { error: roleError } = await supabase
    .from('user_roles')
    .insert([userRoleData]);

  if (roleError) {
    console.error('Error assigning collector role:', roleError);
    return null;
  }

  console.log('Collector account created successfully:', staffDataResult?.[0]);
  return authData.user;
}

async function createCreditorAccount(email: string, fullName: string) {
  console.log(`Creating creditor account for ${fullName}...`);
  
  // First, create the user account
  const { data: authData, error: authError } = await supabase.auth.signUp({
    email: email,
    password: PASSWORD,
    options: {
      data: {
        full_name: fullName,
      }
    }
  });

  if (authError) {
    console.error('Error creating creditor auth account:', authError);
    return null;
  }

  if (!authData || !authData.user) {
    console.error('Failed to create creditor auth account - no user data returned');
    return null;
  }

  console.log('Creditor auth account created:', authData.user.id);
  
  // Create profile record first
  const profileData = {
    id: authData.user.id,
    full_name: fullName,
    email: email,
    updated_at: new Date().toISOString(),
  };

  const { error: profileError } = await supabase
    .from('profiles')
    .upsert([profileData], { onConflict: 'id' });

  if (profileError) {
    console.error('Error creating profile record:', profileError);
    return null;
  }

  // Generate unique employee ID
  const employeeId = `CREDITOR${Math.floor(Math.random() * 10000)}`;
  
  // Create staff record
  const staffData = {
    user_id: authData.user.id,
    employee_id: employeeId,
  };

  const { data: staffDataResult, error: staffError } = await supabase
    .from('staff')
    .insert([staffData])
    .select();

  if (staffError) {
    console.error('Error creating staff record:', staffError);
    return null;
  }

  // Assign creditor role
  const userRoleData = {
    user_id: authData.user.id,
    role: 'creditor',
    active: true,
  };

  const { error: roleError } = await supabase
    .from('user_roles')
    .insert([userRoleData]);

  if (roleError) {
    console.error('Error assigning creditor role:', roleError);
    return null;
  }

  console.log('Creditor account created successfully:', staffDataResult?.[0]);
  return authData.user;
}

async function setupCollectorCreditorAccounts() {
  console.log('Setting up collector and creditor accounts...');
  
  try {
    // Create collector account
    const collectorUser = await createCollectorAccount(
      'collector@example.com', 
      'John Collector'
    );
    
    // Add delay to avoid rate limiting
    await delay(2000);
    
    // Create creditor account
    const creditorUser = await createCreditorAccount(
      'creditor@example.com', 
      'Jane Creditor'
    );
    
    console.log('\n=== ACCOUNT SETUP COMPLETE ===');
    console.log('Collector Account:');
    console.log('  Email: collector@example.com');
    console.log('  Password: Test1234!');
    console.log('\nCreditor Account:');
    console.log('  Email: creditor@example.com');
    console.log('  Password: Test1234!');
    console.log('\nAll accounts have been set up successfully!');
    
  } catch (error) {
    console.error('Error setting up collector/creditor accounts:', error);
  }
}

// Run the setup if this script is executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  setupCollectorCreditorAccounts().catch(console.error);
}

export { createCollectorAccount, createCreditorAccount };